tests/lib/linter.js

```diff
@@ -0,0 +1,74 @@
+/**
+ * @fileoverview Test file for Linter class
+ * @author Gyandeep Singh
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+const assert = require("chai").assert,
+    Linter = require("../../lib/linter");
+
+//------------------------------------------------------------------------------
+// Tests
+// All the core logic has been tested inside eslint.js file
+// Here we will be focusing more on the mutability portion of it
+//------------------------------------------------------------------------------
+
+/**
+ * extract the keys into an array
+ * @param {Map} mapObj - Map type object
+ * @returns {Array<*>} collection of all the keys
+ * @private
+ */
+function extractMapKeys(mapObj) {
+    const keys = [];
+
+    for (const key of mapObj.keys()) {
+        keys.push(key);
+    }
+
+    return keys;
+}
+
+describe("Linter", () => {
+    describe("mutability", () => {
+        let linter1 = null;
+        let linter2 = null;
+
+        beforeEach(() => {
+            linter1 = new Linter();
+            linter2 = new Linter();
+        });
+
+        describe("rules", () => {
+            it("with no changes, same rules are loaded", () => {
+                assert.sameDeepMembers(extractMapKeys(linter1.getRules()), extractMapKeys(linter2.getRules()));
+            });
+
+            it("loading rule in one doesnt change the other", () => {
+                linter1.defineRule("mock-rule", () => ({}));
+
+                assert.isTrue(linter1.getRules().has("mock-rule"), "mock rule is present");
+                assert.isFalse(linter2.getRules().has("mock-rule"), "mock rule is not present");
+            });
+        });
+
+        describe("environments", () => {
+            it("with no changes same env are loaded", () => {
+                // eslint-disable-next-line no-underscore-dangle
+                assert.sameDeepMembers(extractMapKeys(linter1.environments._environments), extractMapKeys(linter2.environments._environments));
```

Comment:
I don't think our tests should depend on the `_environments` property, because it's a private implementation detail and could change in the future. Is there another way to test this?